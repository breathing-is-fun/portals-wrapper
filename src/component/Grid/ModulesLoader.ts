// import { fetch } from 'whatwg-fetch';

const importPolyfill = (url: string) => {
  if (!url) {
    console.error('importPolyfill error: url is undefined.');
    return;
  }

  const promise = fetch(url)
    .then(reponse => reponse.text())
    .then(text => {
      const module = { exports: {} };
      const generate = new Function('module', text);

      generate(module);

      return module.exports;
    })
    .catch(error => console.error('importPolyfill error: ' + error));

  return promise;
};

export default class ModulesLoader {
  private layout: Array<any>;
  private roots: any;

  constructor(layout: Array<any>, roots: any) {
    this.layout = layout;
    this.roots = roots;
  }

  load = (tool: any) => {
    let pathArr = [],
      newLayout = [];

    for (let item of this.layout) {
      const { path, moduletype: type } = item;

      if (type != 'iframe') {
        pathArr.push(importPolyfill(path));
        newLayout.push(item);
      }
    }

    this.layout = newLayout;
    this.loadScripts(pathArr, 0, [], (modules: Array<any>) => {
      for (let i = 0; i < modules.length; i++) {
        let TargetModule = modules[i];

        if (!TargetModule) {
          continue;
        }

        try {
          const { i: key } = this.layout[i];

          if ('default' in TargetModule) {
            TargetModule = TargetModule.default;
          }

          const loadedModule = new TargetModule(this.roots[key], tool);
          const { _moduleOnMount } = loadedModule;

          _moduleOnMount && _moduleOnMount.call(loadedModule);
        } catch (error) {
          console.warn('Ignored error => ' + error);
        }
      }
    });
  };

  loadScripts = (
    pathArr: Array<any>,
    index = 0,
    importedModules: Array<any> = [],
    callback: (importedModules: Array<any>) => void,
  ) => {
    if (index != pathArr.length) {
      pathArr[index]
        .then((importModule: any) => {
          importedModules.push(importModule);
          this.loadScripts(pathArr, ++index, importedModules, callback);
        })
        .catch((error: any) => console.error('loadScripts error: ' + error));
    } else {
      callback && callback(importedModules);
    }
  };
}
