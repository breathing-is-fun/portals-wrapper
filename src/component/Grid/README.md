English | [简体中文](./README-zh_CN.md)

# 简介

- 每个第三方插件的容器，处理布局及布局的保存

- 接收拖拽结束的容器

- 受控

- 基于 [react-grid-layout](https://github.com/STRML/react-grid-layout)
  
# API  

`Grid` 中包含很多组件，像外壳 `Shell`、属性面板 `PropertyBoard`，`Grid` 更多的时候被当做入口
  
换句话说，其他组件的属性在这里也可以用，`Grid` 仅作传递

## Grid的属性

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| layout | 布局数据 | [[object](#layout对象说明)] | - |
| onLayoutChange | 布局改变后的回调 | function (newLayout: []) | - |

## Shell中的属性

`isEdit`、`isDelete`、`onDelete`、`onDetail`

## PropertyBoard中的属性

`enumDatas`

## layout对象说明

除了 [react-grid-layout](https://github.com/STRML/react-grid-layout#grid-layout-props)中自带的属性，你还需要一些其他属性更好地控制[插件](http://www.baidu.com)的显示
| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| type | 插件类型，暂时只有一种枚举 | string, enum['iframe'] | - |
| path | 插件文件地址，需要指向入口文件，[入口文件需要像这样才行](http://www.baidu.com)；如果是 `type == iframe`，则需要传 `url` 地址 | string, url | - |
| style | 控制 `Shell` 最外层的样式，比如阴影啊边框什么的。需要能用 `JSON.parse` 转成对象的字符串 | string, json | - |
| i | 用作插件的唯一标识，用这个关联数据和实际节点 | `string|number` | - |

注：这里的 `path`，不论如何都需要同源，这是因为[同源](https://developer.mozilla.org/zh-CN/docs/Web/Security/Same-origin_policy)限制了浏览器行为

# 方法

| 名称 | 说明 |
| --- | --- |
| mountRoots | 封装了 `ModulesLoader`，用作加载插件，[了解更多](#ModulesLoader) |

## ModulesLoader

``` javascript
    import Loader from './ModulesLoader';

    /**
     * @param layout 布局数据
     * 可以参考Grid中的layout，这玩意跟它一样
     * @param roots 组件挂载的节点
     * { test: DOM }，传进来需要像这样
     * 到时候会把字段名也就是test当作节点的唯一标识
     * 值需要是实际的DOM
    */
    const loader = new Loader(layout, roots);

    /**
     * 上面是初始化，这里才开始加载
    */
    loader.load();
```

# 示例

![img](https://camo.githubusercontent.com/8c68a2e6d6e01364247232267a5698ac0d9b63c6/687474703a2f2f692e696d6775722e636f6d2f6f6f314e5436632e676966)
  
直接抄人家的图用来说明格子的概念