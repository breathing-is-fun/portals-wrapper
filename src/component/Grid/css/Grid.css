.Grid iframe {
    border: none;
    width: 100%;
    overflow: hidden;
}

.Grid .render-div {
    overflow: auto;
    padding: 15px;
}