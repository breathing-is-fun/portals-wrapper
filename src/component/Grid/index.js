import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Responsive, WidthProvider } from 'react-grid-layout';
const ResponsiveGridLayout = WidthProvider(Responsive);

import 'react-grid-layout/css/styles.css';
import 'react-resizable/css/styles.css';

import Shell from '../../component/Shell';
import Loader from './ModulesLoader';

import './css/Grid.css';

export default class Grid extends Component {
	constructor (props) {
		super(props);

		this.state = {
			isDrawerOpen: false,
			propertyBoardDataSource: {},
			PropertyBoard: null,
			rowHeight: 30,
		};

		this.roots = {};
	}

	static defaultProps = {
		showEdit: false,
		showDelete: false,
		layout: [],
		enumDatas: [],
	}

	mountRoots = () => {
		const { layout } = this.props;

    	const loader = new Loader(layout, this.roots);

    	loader.load();
	}

	handleLayoutChange = layout => {
		const { onLayoutChange, layout: propsLayout } = this.props;

		let newLayout = [];

		for(let i = 0; i < layout.length; i++) {
			const item = layout[i];
			const jtem = propsLayout[i];

			newLayout.push(Object.assign({}, jtem, item));
		}

		onLayoutChange && onLayoutChange(newLayout);
	}

	handleDragDrop = e => {
		let { layout, onLayoutChange } = this.props;
		const item = JSON.parse(e.dataTransfer.getData('menuItemToGrid'));
		const { key, imgurl: imgUrl, text, style, url } = item;
		const defaultProps = {
			i: '' + key + layout.length,
			x: (layout.length * 2) % 12,
			w: 2, h: 9, y: 0,
			url,
			path: url,
			imgUrl,
			title: text,
			style,
		};

		layout.push(Object.assign({}, item, defaultProps));

		onLayoutChange && onLayoutChange(layout);
	}

	createShellChild = (showEdit, item) => {
		const { i, moduletype: type, imgurl, imgUrl, path } = item;
		const height = 'calc(100% - 21px)';

		if(showEdit) {
			return (
				<img
					src={ imgurl || imgUrl }
					style={{ width: '100%', height }}
				/>
			);
		}

		if(type == 'iframe') {
			return (
				<iframe
					src={ path }
					style={{ height }}
				></iframe>
			);
		}

		return (
			<div
				className='render-div'
				style={{ height }}
				ref={ ref => ref && (this.roots[i] = ref) }
				id={ i }
			/>
		);
	}

	handleShellonEdit = item => {
		import('../PropertyBoard')
			.then(PropertyBoard => {
				this.setState({
					isDrawerOpen: true,
					propertyBoardDataSource: item,
					PropertyBoard: PropertyBoard.default || PropertyBoard,
				});
			});
	}

	handleShellStyleOnChange = currentShellStyle => {
		const { onLayoutChange, layout } = this.props;
		let newLayout = [];

		for(let item of layout) {
			const { i: key, style = {} } = item;
			const { id } = currentShellStyle;
			let obj = item;

			if(key == id) {
				obj.style = Object.assign(style, currentShellStyle);
				delete obj.style.id;
			}

			newLayout.push(obj);
		}

		onLayoutChange && onLayoutChange(newLayout);
	}

	handleBreakpointChange = (breakpoints, cols) => {
		this.setState({ rowHeight: cols * 2.5 });
	}

    render = () => {
    	const {
    		isDrawerOpen,
    		propertyBoardDataSource,
    		PropertyBoard,
    		rowHeight,
    	} = this.state;
    	const {
    		showEdit,
    		showDelete,
    		layout,
    		onDelete,
    		onDetail,
    		enumDatas
    	} = this.props;

    	const layoutProps = {
    		className: 'layout',
    		cols: { lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 },
    		breakpoints: { lg: 1300, md: 996, sm: 768, xs: 480, xxs: 0 },
    		rowHeight,
    		width: document.body.clientWidth - (showEdit ? 256 : 0),
    		// margin: [10, 10],
    		onLayoutChange: this.handleLayoutChange,
    		isDraggable: showEdit,
    		isResizable: showEdit,
    		// compactType: 'horizontal',
    		style: {
    			background: '#f5f6fa',
    			height: layout.length != 0 ? '100%' : 500
    		},
    		onBreakpointChange: this.handleBreakpointChange,
    	};

    	const propertyBoardProps = {
    		visible: isDrawerOpen,
    		onClose: isDrawerOpen => {
    			this.setState({ isDrawerOpen }, () => {
    				// 关闭抽屉时销毁外壳中的元素，好在再次点击时执行componentDidMount中的方法
    				setTimeout(() => {
    					this.setState({ PropertyBoard: null });
    				}, 300);
    			});
    		},
    		shellStyleDatas: propertyBoardDataSource,
    		enumDatas,
    		onChange: this.handleShellStyleOnChange,
    	};

    	return (
    		<div
    			className='Grid'
    			onDrop={ this.handleDragDrop }
    			onDragOver={ e => e.preventDefault() }
    		>
    			<ResponsiveGridLayout { ...layoutProps }>
    				{
    					layout.map(item => {
    						const {
    							i: key,
    							title,
    							style = {},
    							showtitle: showTitle,
    							showdetail: showDetail,
    							detailpath: detailPath,
    						} = item;

    						const shellProps = {
    							key,
    							title,
    							showEdit,
    							showDelete,
    							onDelete,
    							showTitle,
    							onDetail,
    							detailPath,
    							showDetail,
    							style,
    							'data-grid': item,
    							onEdit: dataGrid => {
    								this.handleShellonEdit(dataGrid);
    							},
    						};

    						return (
    							<Shell { ...shellProps }>
    								{ this.createShellChild(showEdit, item) }
    							</Shell>
    						);
    					})
    				}
    			</ResponsiveGridLayout>

    			{ PropertyBoard && <PropertyBoard { ...propertyBoardProps } /> }
    		</div>
    	);
    }
}

Grid.propTypes = {
	layout: PropTypes.array,
	onLayoutChange: PropTypes.func,
	enumDatas: PropTypes.array,
	showEdit: PropTypes.bool,
	showDelete: PropTypes.bool,
	onDelete: PropTypes.func,
	onDetail: PropTypes.func,
};