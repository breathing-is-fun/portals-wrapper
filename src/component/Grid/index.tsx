import React, { Component } from 'react';
import { Responsive, WidthProvider } from 'react-grid-layout';
const ResponsiveGridLayout = WidthProvider(Responsive);
import 'react-grid-layout/css/styles.css';
import 'react-resizable/css/styles.css';
import Shell from '../../component/Shell';
import Loader from './ModulesLoader';
import './css/Grid.css';

export interface GridProps {
  layout: Array<any>;
  enumDatas: Array<string>;
  showEdit?: boolean;
  showDelete?: boolean;
  onAdd?: (dataGrid: any) => void;
  onDelete?: (dataGrid: any) => void;
  onDetail?: (dataGrid: any) => void;
  onEdit?: (dataGrid: any) => void;
  onLayoutChange?: (newLayout: Array<any>) => void;
  size?: string | number;
  isResizable?: boolean;
  isDraggable?: boolean;
}
export interface GridState {
  isDrawerOpen: boolean;
  propertyBoardDataSource: any;
  PropertyBoard: any;
  rowHeight: number;
}

export default class Grid extends Component<GridProps, GridState> {
  private roots: any;
  static defaultProps = {
    layout: [],
    enumDatas: [],
  };

  constructor(props: GridProps) {
    super(props);
    this.state = {
      isDrawerOpen: false,
      propertyBoardDataSource: {},
      PropertyBoard: null,
      rowHeight: 30,
    };
    this.roots = {};
  }

  mountRoots = () => {
    const { layout } = this.props;

    const loader = new Loader(layout, this.roots);

    loader.load(SCTool);
  };

  handleLayoutChange = (layout: Array<any>) => {
    const { onLayoutChange, layout: propsLayout } = this.props;

    let newLayout = [];

    for (let i = 0; i < layout.length; i++) {
      const item = layout[i];
      const jtem = propsLayout[i];

      newLayout.push(Object.assign({}, jtem, item));
    }

    onLayoutChange && onLayoutChange(newLayout);
  };

  handleDragDrop = (e: any) => {
    let { layout, onLayoutChange } = this.props;
    const item = JSON.parse(e.dataTransfer.getData('menuItemToGrid'));
    const { key, imgurl: imgUrl, text, style, url } = item;
    const defaultProps = {
      i: '' + key + layout.length,
      x: (layout.length * 2) % 12,
      w: 2,
      h: 9,
      y: 0,
      url,
      path: url,
      imgUrl,
      title: text,
      style,
    };

    layout.push(Object.assign({}, item, defaultProps));

    onLayoutChange && onLayoutChange(layout);
  };

  createShellChild = (showEdit: boolean, item: any) => {
    const { i, moduletype: type, imgurl, imgUrl, path } = item;
    const height = 'calc(100% - 33px)';

    if (showEdit) {
      return <img src={imgurl || imgUrl} style={{ width: '100%', height }} />;
    }

    if (type == 'iframe') {
      return <iframe src={path} style={{ height }} />;
    }

    return (
      <div
        className="render-div"
        style={{ height }}
        ref={ref => ref && (this.roots[i] = ref)}
        id={i}
      />
    );
  };

  handleShellonEdit = (item: any) => {
    import('../PropertyBoard').then(PropertyBoard => {
      this.setState({
        isDrawerOpen: true,
        propertyBoardDataSource: item,
        PropertyBoard: PropertyBoard.default || PropertyBoard,
      });
    });
  };

  handleShellStyleOnChange = (currentShellStyle: any) => {
    const { onLayoutChange, layout } = this.props;
    let newLayout = [];

    for (let item of layout) {
      const { i: key, style = {} } = item;
      const { id } = currentShellStyle;
      let obj = item;

      if (key == id) {
        obj.style = Object.assign(style, currentShellStyle);
        delete obj.style.id;
      }

      newLayout.push(obj);
    }

    onLayoutChange && onLayoutChange(newLayout);
  };

  handleBreakpointChange = (breakpoints: string | number, cols: number) => {
    this.setState({ rowHeight: cols * 2.5 });
  };

  render = () => {
    const {
      isDrawerOpen,
      propertyBoardDataSource,
      PropertyBoard,
      rowHeight,
    } = this.state;
    const {
      showEdit,
      showDelete,
      layout,
      onDelete,
      onDetail,
      enumDatas,
      size,
      onAdd,
      isDraggable,
      isResizable,
    } = this.props;

    const layoutProps = {
      className: 'layout',
      cols: { lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 },
      breakpoints: { lg: 1270, md: 996, sm: 768, xs: 480, xxs: 0 },
      rowHeight,
      width: document.body.clientWidth - (showEdit ? 256 : 0),
      margin: [20, 25],
      onLayoutChange: this.handleLayoutChange,
      isDraggable: showEdit || isDraggable,
      isResizable: showEdit || isResizable,
      // compactType: 'horizontal',
      style: {
        background: '#f5f6fa',
        height: layout.length != 0 ? '100%' : 500,
      },
      onBreakpointChange: this.handleBreakpointChange,
    };

    const propertyBoardProps = {
      visible: isDrawerOpen,
      onClose: (isDrawerOpen: boolean) => {
        this.setState({ isDrawerOpen }, () => {
          // 关闭抽屉时销毁外壳中的元素，好在再次点击时执行componentDidMount中的方法
          setTimeout(() => {
            this.setState({ PropertyBoard: null });
          }, 300);
        });
      },
      shellStyleDatas: propertyBoardDataSource,
      enumDatas,
      onChange: this.handleShellStyleOnChange,
    };

    return (
      <div
        className="Grid"
        onDrop={this.handleDragDrop}
        onDragOver={e => e.preventDefault()}
      >
        <ResponsiveGridLayout {...layoutProps}>
          {layout.map(item => {
            let {
              i: key,
              title,
              style = {},
              showtitle: showTitle,
              showdetail: showDetail,
              detailpath: detailPath,
            } = item;

            if (typeof style == 'string') {
              style = JSON.parse(style);
            }

            const shellProps = {
              key,
              title,
              showEdit,
              showDelete,
              onDelete,
              showTitle,
              onDetail,
              detailPath,
              showDetail,
              style,
              size,
              onAdd,
              'data-grid': item,
              onEdit: (dataGrid: any) => {
                this.handleShellonEdit(dataGrid);
              },
            };

            return (
              <Shell {...shellProps}>
                {this.createShellChild(!!showEdit, item)}
              </Shell>
            );
          })}
        </ResponsiveGridLayout>

        {PropertyBoard && <PropertyBoard {...propertyBoardProps} />}
      </div>
    );
  };
}
