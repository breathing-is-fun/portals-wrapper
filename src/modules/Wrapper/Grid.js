/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-09-26 11:25:50
 * @Last Modified by: zy9
 * @Last Modified time: 2018-09-29 15:46:01
 */
import React, { Component } from 'react';

import GridLayout from 'react-grid-layout';
import 'react-grid-layout/css/styles.css';
import 'react-resizable/css/styles.css';

export default class Grid extends Component {
	constructor (props) {
		super(props);

		this.state = {
			layout: [],
		};
	}

    componentDidMount = () => {
    	this.loadLayout(() => {
    		const { layout } = this.state;
    		let pathArr = [];

    		for(let item of layout) {
    			const { path, type } = item;

    			type != 'iframe' && pathArr.push(`import('${ path }')`);
    		}
    		pathArr = `[${ pathArr.toString() }]`;

    		/* eslint-disable no-eval */
    		Promise.all(eval(`${ pathArr }`)).then(modules => {
    			for(let i = 0; i < modules.length; i++) {
    				const { TestModule } = modules[i];
    				const { i: key } = layout[i];
    				const testModule = new TestModule(this[key]);
    				const { _moduleOnMount } = testModule;

    				_moduleOnMount && _moduleOnMount.call(testModule);
    			}
    		});
    	});
    }

	loadLayout = callback => {
		fetch('../../../mock/layoutDatas.json')
			.then(result => result.json())
			.then(result => {
				const { layout } = result;

				this.setState({ layout }, () => callback && callback());
			});
	}

    handleLayoutChange = layout => {
    	console.log(layout);
    }

	handleDragDrop = e => {
		let { layout } = this.state;
		const item = JSON.parse(e.dataTransfer.getData('menuItemToGrid'));
		const { key, url, text } = item;

		layout.push({
			i: '' + key + layout.length,
			x: (layout.length * 2) % 12,
			y: Infinity,
			w: 2,
			h: 9,
			type: 'iframe',
			imgUrl: url,
			title: text
		});

		this.setState({ layout });
	}

    render = () => {
    	const { layout } = this.state;

    	const layoutProps = {
    		className: 'layout',
    		// layout,
    		cols: 12,
    		rowHeight: 30,
    		width: (document.documentElement.clientWidth || document.body.clientWidth) - 256,
    		margin: [10, 10],
    		onLayoutChange: this.handleLayoutChange,
    	};

    	return (
    		<div className='Grid' onDrop={ this.handleDragDrop } onDragOver={ e => e.preventDefault() }>
    			<GridLayout { ...layoutProps }>
    				{
    					layout.map(item => {
    						const { i, type, imgUrl, title } = item;

    						return type == 'iframe' ? (
    							<div key={ i } data-grid={ item } style={{ zIndex: 1 }}>
    								<div style={{ height: 30, background: '#F96' }}>{ title }</div>
    								<img src={ imgUrl } style={{ width: '100%', height: 'calc(100% - 30px)' }} />
    							</div>
    						) : (
    							<div key={ i } data-grid={ item } style={{ background: '#ccc' }} ref={ ref => this[i] = ref } />
    						);
    					})
    				}
    			</GridLayout>
    		</div>
    	);
    }
}