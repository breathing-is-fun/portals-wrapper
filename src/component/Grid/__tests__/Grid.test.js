import React from 'react';
import { render, mount, shallow } from 'enzyme';
import Grid from '..';

describe('Grid', () => {
  const layoutItem = {
    h: 6,
    i: 'dbzx',
    id: 1,
    imgurl: 'http://47.95.1.229:9003/UploadFile/201808/dbzx.png',
    mealid: 1,
    path: '../../../thirdModules/dbzx/index.js',
    showtitle: false,
    style: { background: '#F96', border: '1px solid #ccc' },
    title: '代办中心',
    w: 4,
    x: 0,
    y: 0,
  };

  it('render correctly', () => {
    const wrapper = mount(<Grid layout={[layoutItem]} />);

    expect(wrapper).toMatchSnapshot();
  });
});
