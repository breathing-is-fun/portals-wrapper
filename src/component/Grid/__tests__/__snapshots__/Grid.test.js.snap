// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid render correctly 1`] = `
<Grid
  isDelete={false}
  isEdit={false}
  layout={
    Array [
      Object {
        "h": 6,
        "i": "dbzx",
        "id": 1,
        "imgurl": "http://47.95.1.229:9003/UploadFile/201808/dbzx.png",
        "mealid": 1,
        "path": "../../../thirdModules/dbzx/index.js",
        "showtitle": false,
        "style": Object {
          "background": "#F96",
          "border": "1px solid #ccc",
        },
        "title": "代办中心",
        "w": 4,
        "x": 0,
        "y": 0,
      },
    ]
  }
  propertyBoardEnumData={Array []}
>
  <div
    className="Grid"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <ReactGridLayout
      autoSize={true}
      className="layout"
      cols={12}
      compactType="vertical"
      containerPadding={null}
      draggableCancel=""
      draggableHandle=""
      isDraggable={false}
      isResizable={false}
      layout={Array []}
      margin={
        Array [
          10,
          10,
        ]
      }
      maxRows={Infinity}
      onDrag={[Function]}
      onDragStart={[Function]}
      onDragStop={[Function]}
      onLayoutChange={[Function]}
      onResize={[Function]}
      onResizeStart={[Function]}
      onResizeStop={[Function]}
      preventCollision={false}
      rowHeight={30}
      style={
        Object {
          "background": "#f5f6fa",
          "height": "100%",
        }
      }
      useCSSTransforms={true}
      verticalCompact={true}
      width={0}
    >
      <div
        className="react-grid-layout layout"
        style={
          Object {
            "background": "#f5f6fa",
            "height": "100%",
          }
        }
      >
        <GridItem
          cancel=""
          className=""
          cols={12}
          containerPadding={
            Array [
              10,
              10,
            ]
          }
          containerWidth={0}
          h={6}
          handle=""
          i="dbzx"
          isDraggable={false}
          isResizable={false}
          key=".$dbzx"
          margin={
            Array [
              10,
              10,
            ]
          }
          maxH={Infinity}
          maxRows={Infinity}
          maxW={Infinity}
          minH={1}
          minW={1}
          onDrag={[Function]}
          onDragStart={[Function]}
          onDragStop={[Function]}
          onResize={[Function]}
          onResizeStart={[Function]}
          onResizeStop={[Function]}
          rowHeight={30}
          static={false}
          useCSSTransforms={true}
          usePercentages={false}
          w={4}
          x={0}
          y={0}
        >
          <Shell
            className="react-grid-item cssTransforms"
            data-grid={
              Object {
                "h": 6,
                "i": "dbzx",
                "id": 1,
                "imgurl": "http://47.95.1.229:9003/UploadFile/201808/dbzx.png",
                "mealid": 1,
                "path": "../../../thirdModules/dbzx/index.js",
                "showtitle": false,
                "style": Object {
                  "background": "#F96",
                  "border": "1px solid #ccc",
                },
                "title": "代办中心",
                "w": 4,
                "x": 0,
                "y": 0,
              }
            }
            detailPath=""
            isDelete={false}
            isEdit={false}
            key="dbzx"
            onEdit={[Function]}
            showDetail={false}
            showTitle={false}
            style={
              Object {
                "MozTransform": "translate(10px,10px)",
                "OTransform": "translate(10px,10px)",
                "WebkitTransform": "translate(10px,10px)",
                "background": "#F96",
                "border": "1px solid #ccc",
                "height": "230px",
                "msTransform": "translate(10px,10px)",
                "position": "absolute",
                "transform": "translate(10px,10px)",
                "width": "-13px",
              }
            }
            title="代办中心"
            type="component"
          >
            <div
              className="Shell"
              data-grid={
                Object {
                  "h": 6,
                  "i": "dbzx",
                  "id": 1,
                  "imgurl": "http://47.95.1.229:9003/UploadFile/201808/dbzx.png",
                  "mealid": 1,
                  "path": "../../../thirdModules/dbzx/index.js",
                  "showtitle": false,
                  "style": Object {
                    "background": "#F96",
                    "border": "1px solid #ccc",
                  },
                  "title": "代办中心",
                  "w": 4,
                  "x": 0,
                  "y": 0,
                }
              }
              onClick={[Function]}
              style={
                Object {
                  "MozTransform": "translate(10px,10px)",
                  "OTransform": "translate(10px,10px)",
                  "WebkitTransform": "translate(10px,10px)",
                  "background": "#F96",
                  "border": "1px solid #ccc",
                  "height": "230px",
                  "msTransform": "translate(10px,10px)",
                  "position": "absolute",
                  "transform": "translate(10px,10px)",
                  "width": "-13px",
                }
              }
              type="component"
            >
              <div
                className="render-div"
                id="dbzx"
                style={
                  Object {
                    "height": "calc(100% - 21px)",
                  }
                }
              />
            </div>
          </Shell>
        </GridItem>
      </div>
    </ReactGridLayout>
  </div>
</Grid>
`;
